docs/user-guide/v3_migration.rst

```diff
@@ -0,0 +1,204 @@
+following actions:
+
+1. Pin the supported Zarr-Python version to ``zarr>=2,<3``. This is a best practice
+   and will protect your users from any incompatibilities that may arise during the
+   release of Zarr-Python 3.0.
+2. Limit your imports from the Zarr-Python package. Most of the primary API ``zarr.*``
+   will be compatible in 3.0. However, the following breaking API changes are planned:
+
+   - ``numcodecs.*`` will no longer be available in ``zarr.*``. To migrate, import codecs
+     directly from ``numcodecs``:
+
+     .. code-block:: python
+
+        from numcodecs import Blosc
+        # instead of:
+        # from zarr import Blosc
+
+   - The ``zarr.v3_api_available`` feature flag is being removed. In Zarr-Python 3.0
+     the v3 API is always available, so you shouldn't need to use this flag.
+   - The following internal modules are being removed or significantly changed. If
+     your application relies on imports from any of the below modules, you will need
+     to either a) modify your application to no longer rely on these imports or b)
+     vendor the parts of the specific modules that you need.
+
+     * ``zarr.attrs``
+     * ``zarr.codecs``
+     * ``zarr.context``
+     * ``zarr.core``
+     * ``zarr.hierarchy``
+     * ``zarr.indexing``
+     * ``zarr.meta``
+     * ``zarr.meta_v1``
+     * ``zarr.storage``
+     * ``zarr.sync``
+     * ``zarr.types``
+     * ``zarr.util``
+     * ``zarr.n5``
+
+3. Test that your package works with v3. You can start testing against version 3 now
+   (pre-releases are being published to PyPI weekly).
+4. Update the pin to zarr >=3
+
+Continue using Zarr-Python 2
+----------------------------
+
+Zarr-Python 2.x is still available, though we recommend migrating to Zarr-Python 3 for
+its improvements and new features. Security and bug fixes will be made to the 2.x series
+for at least 6 months following the first Zarr-Python 3 release.
+If you need to use the latest Zarr-Python 2 release, you can install it with:
+
+.. code-block:: console
+
+    $ pip install "zarr==2.*"
+
+Migration Guide
+---------------
+
+The following sections provide details on the most important changes in Zarr-Python 3.
+
+The Array class
+~~~~~~~~~~~~~~~
+
+1. Disallow direct construction - use :func:`zarr.open_array` or :func:`zarr.create_array`
+   instead of directly constructing the :class:`zarr.Array` class.
+
+2. Defaulting to ``zarr_format=3`` - newly created arrays will use the version 3 of the
+   Zarr specification. To continue using version 2, set ``zarr_format=2`` when creating arrays
+   or set ``default_zarr_version=2`` in :ref:`config`.
+
+The Group class
+~~~~~~~~~~~~~~~
+
+1. Disallow direct construction - use :func:`zarr.open_group` or :func:`zarr.create_group`
+   instead of directly constructing the :class:`zarr.Group` class.
+2. Deprecated most of the h5py compatibility methods. The following migration is suggested:
```

Comment:
Are the suggested functions drop in replacements with the same arguments? If not, can we provide a clearer guide on how to translate existing calls to create_dataset/require_dataset to create_array/require_array?